#!/usr/bin/env python
import vcf
from Bio import SeqIO
import math
import random
import argparse

USAGE = """
simulate_fastq.py --vcf <snps VCF file>
                  --bed <GBSBED file of restriction sites>
                  --fasta <reference fasta>
"""

DESCRIPTION = """
Simulate single-end sequencing reads from a GBS experiment. Reads are
drawn from a fasta file (ALT alleles are drawn randomly). Sample genotypes
are taken from the input VCF file. The location of the reads is taken from
the GBSBED file (generated by make_gbsbed.py). The locations in the VCF
file do not correspond to those in the GBSBED file; only the genotypes from
the VCF are used.
"""

parser = argparse.ArgumentParser(usage=USAGE, description=DESCRIPTION)
parser.add_argument('--vcf', dest='vcf', help='snps vcf file')
parser.add_argument('--bed', dest='bed', help='GBSBED file of restriction sites')
parser.add_argument('--fasta', dest='fasta', help='reference genomoe fasta')
parser.add_argument('--seed', dest='seed', type=int, default=0, help='seed for random number generator')
parser.add_argument('--readlen', dest='readlen', type=int, default=101, help='read length (default=101)')
parser.add_argument('--fragment_max', dest='fragment_max', type=int, default=500, help='maximum length of restriction fragments (default=500)')
parser.add_argument('--epsilon', dest='epsilon', type=float, default=0.001, help='base call error rate (default=0.001)')
args = parser.parse_args()

rand = random.Random(args.seed)
readlen = args.readlen
maxlen = args.fragment_max
epsilon = args.epsilon
def write_reads(seq, counts, myfile):
    '''Write sequence to fastq file.'''
    for i in range(counts):
        lane = rand.randint(1,8)
        tile = rand.randint(1,100)
        x = rand.randint(1,9999)
        y = rand.randint(1,9999)
        seqid = "@SIMULATION:%i:%i:%i:%i#0" % (lane, tile, x, y)
        qual = [chr(int(-10 * math.log(epsilon, 10) + 33))] * len(seq)
        qual = ''.join(qual)
        myfile.write(seqid + '\n')
        myfile.write(seq + '\n')
        myfile.write('+\n')
        myfile.write(qual + '\n')
        
# Get the reference sequence.
seq_record = SeqIO.parse(open(args.fasta, 'r'), 'fasta').next()

# Get a list of reads from the in silico digest of the reference.
reads = []
bed = open(args.bed, 'r')
for line in bed:
    line = line.strip()
    if line[0] == "#":
        continue
    (chrom, start, end, fwd_insert, rev_insert, enzyme, strand, 
     site_id, freq, cut_allele, fwd_lig, rev_lig) = line.split()
    if fwd_insert != '.' and rev_insert != '.':
        fwd_insert = int(fwd_insert)
        rev_insert = int(rev_insert)
        fwd_lig = max([int(i) for i in fwd_lig.split(',')])
        rev_lig = min([int(i) for i in rev_lig.split(',')])
        if fwd_insert >= 2 * readlen and fwd_insert <= maxlen:
            read = seq_record.seq[fwd_lig:fwd_lig + readlen]
            reads.append(str(read))
        if rev_insert >= 2 * readlen and rev_insert <= maxlen:
            read = seq_record.seq[rev_lig - readlen + 1:rev_lig + 1]
            read = read.reverse_complement()
            reads.append(str(read))
    else:
        pass
readiter = (read for read in reads)

# Make a hash of fastq files, keyed by sample name.
fastq = {}
reader = vcf.Reader(open(args.vcf, 'r'))
for sample in reader.samples:
    fastq[sample] = open(sample + '.fastq', 'w')

# Parse the vcf file and write reads for each sample.
for snp in reader:
    ref = readiter.next()
    alt = [base for base in ref]
    bases = ['A', 'G', 'T', 'C']
    bases.remove(ref[50])
    alt[50] = rand.choice(bases)
    alt = ''.join(alt)
    for sample in snp.samples:
        if sample['AD']:
            write_reads(ref, int(sample['AD'][0]), fastq[sample.sample])
            write_reads(alt, int(sample['AD'][1]), fastq[sample.sample])
        else:
            pass

for sample in fastq:
    fastq[sample].close()
    
